feat: add C implementation for `stats/base/dists/triangular/cdf`

0PrashantYadav0 opened the pull request:
Resolves #4668

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for @stdlib/stats/base/dists/triangular/cdf along with relevant tests, benchmarks and examples.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #4668

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

0PrashantYadav0 committed:
Merge branch 'develop' of https://github.com/0PrashantYadav0/stdlib into develop

0PrashantYadav0 committed:
feat: add C implementation for triangular cdf

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 commented:
@Planeshifter please review and merge this pr.

Planeshifter committed:
style: modernize PRNG usage in `triangular/cdf` native benchmark

- Replace individual `randu()` calls with array generation using `uniform()`
- Add proper triangular parameter constraint enforcement (min <= mode <= max)
- Follows same pattern as other modernized benchmarks

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter committed:
fix: simplify example and benchmark in `triangular/cdf`

- Remove constraint loops from example, use safe ranges
- Move random generation outside benchmark timing loop
- Use direct uniform calls for parameter constraints

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/cdf/src/addon.c:
```diff
@@ -0,0 +1,23 @@
+#include "stdlib/stats/base/dists/triangular/cdf.h"
+#include "stdlib/math/base/napi/quaternary.h"
+
+// cppcheck-suppress shadowFunction
```
```suggestion
```

Planeshifter committed:
docs: fix C API documentation in `triangular/cdf` README

- Fix include header path from geometric to triangular
- Fix function names from geometric to triangular throughout C section
- Add missing closing section tags for proper nesting

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter committed:
fix: address code quality issues in `triangular/cdf`

- Fix incorrect function name in C JSDoc example (geometric → triangular)
- Fix C code spacing style (missing space after if)
- Fix JavaScript literal consistency (1 → 1.0)
- Fix test factory call with missing mode parameter

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/cdf/README.md:
```diff
@@ -148,6 +148,117 @@ for ( i = 0; i < 25; i++ ) {
+
+The function accepts the following arguments:
+
+-   **x**: `[in] double` input value.
+-   **a**: `[in] double` lower limit
+-   **b**: `[in] double` upper limit
+-   **c**: `[in] double` mode
```
```suggestion
-   **x**: `[in] double` input value.
-   **a**: `[in] double` lower limit.
-   **b**: `[in] double` upper limit.
-   **c**: `[in] double` mode.
```

Planeshifter committed:
fix: revert example to match README and align native benchmark style

- Revert examples/index.js to match README exactly (randu-based)
- Fix benchmark.native.js to follow benchmark.js style with individual uniform calls
- Use proper Float64Array allocation and individual parameter generation

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter committed:
docs: improve function description in `triangular/cdf` header

- Add more descriptive comment specifying parameter meanings
- Clarify that it evaluates CDF for triangular distribution
- Follow stdlib header documentation conventions

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/cdf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,145 @@
+	for ( i = 0; i < 100; i++ ) {
+		x[ i ] = random_uniform( 0.0, 30.0 );
+		a[ i ] = random_uniform( 0.0, 10.0 );
+		b[ i ] = random_uniform( a[ i ], 40.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
```suggestion
		b[ i ] = random_uniform( a[ i ] + STDLIB_CONSTANT_FLOAT64_EPS, 40.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/cdf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,145 @@
+		x[ i ] = random_uniform( 0.0, 30.0 );
+		a[ i ] = random_uniform( 0.0, 10.0 );
+		b[ i ] = random_uniform( a[ i ], 40.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
+		c[ i ] = a[ i ] + random_uniform( 0.0, b[ i ] - a[ i ] );
```
```suggestion
		c[ i ] = random_uniform( a[ i ], b[ i ] );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/cdf/examples/c/example.c:
```diff
@@ -0,0 +1,46 @@
+	for ( i = 0; i < 25; i++ ) {
+		x = random_uniform( 0.0, 30.0 );
+		a = random_uniform( 0.0, 10.0 );
+		b = random_uniform( a, 40.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
```suggestion
		b = random_uniform( a + STDLIB_CONSTANT_FLOAT64_EPS, 40.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/cdf/examples/c/example.c:
```diff
@@ -0,0 +1,46 @@
+		x = random_uniform( 0.0, 30.0 );
+		a = random_uniform( 0.0, 10.0 );
+		b = random_uniform( a, 40.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
+		c = a + random_uniform( 0.0, b - a );
```
```suggestion
		c = random_uniform( a, b );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/cdf/manifest.json:
```diff
@@ -0,0 +1,84 @@
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/napi/quaternary",
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/constants/float64/ninf",
+        "@stdlib/math/base/special/pow"
```
```suggestion
        "@stdlib/math/base/assert/is-nan",
        "@stdlib/math/base/special/pow"
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/cdf/manifest.json:
```diff
@@ -0,0 +1,84 @@
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/constants/float64/eps",
+        "@stdlib/constants/float64/ninf",
+        "@stdlib/math/base/special/pow"
```
```suggestion
        "@stdlib/constants/float64/eps",
        "@stdlib/math/base/special/pow"
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/cdf/manifest.json:
```diff
@@ -0,0 +1,84 @@
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/constants/float64/eps",
+        "@stdlib/constants/float64/ninf",
+        "@stdlib/math/base/special/pow"
```
```suggestion
        "@stdlib/math/base/assert/is-nan",
        "@stdlib/constants/float64/eps",
        "@stdlib/math/base/special/pow"
```

Planeshifter committed:
chore: minor clean-up

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/cdf/README.md:
```diff
@@ -148,6 +148,117 @@ for ( i = 0; i < 25; i++ ) {
 
 <!-- /.examples -->
 
+<!-- Section to include cited references. If references are included, add a horizontal rule *before* the section. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->
+
+<section class="references">
```
```suggestion
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/cdf/README.md:
```diff
@@ -148,6 +148,117 @@ for ( i = 0; i < 25; i++ ) {
+
+<!-- /.c -->
+
+</section>
+
+<!-- /.references -->
+
```
```suggestion
```

Planeshifter committed:
chore: fix section tags

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter reviewed (approved):
Let's get it in. 🚀
